### PR TITLE
Add mbart-large-cc25, support translation finetuning

### LINKED ISSUE #3513: Adding mbart-large-cc25
# 🌟 New model addition
Multilingual BART model implemented in fairseq introduced by FAIR

## Model description
This issue is to request adding mBART model existing as a part of fairseq lib. 

[Link to the fairseq description of the model](https://github.com/pytorch/fairseq/tree/master/examples/mbart
)
[Link to the mBART paper](https://arxiv.org/abs/2001.08210)

Multilingually pretrained BART checkpoint.

The model code follows the original BART model code which is already a part of ```transformers``` repo. However, it introduces a couple more features like multilingual denoising and translation from pretrained BART. 

## Open source status

- [x] _the model implementation is available: (give details)_
[Link to the PR adding mBART to the fairseq](https://github.com/pytorch/fairseq/commit/5e79322b3a4a9e9a11525377d3dda7ac520b921c) 
This PR shows the main pieces that were added to the fairseq to make mBART work considering BART  which is already existing in the codebase. However, a few additional mBART commits were added afterward.

- [x] _the model weights are available: (give details)_
[Link to the weights](https://github.com/pytorch/fairseq/tree/master/examples/mbart#pre-trained-models)

- [x] _who are the authors: (mention them, if possible by @gh-username)_
Facebook AI Research (@MultiPath)

### LINKED ISSUE #5458: 🐛 Can't use `AutoTokenizer` with `sshleifer/mbart-large-cc25`
# 🐛 Bug

From [`sshleifer/mbart-large-cc25`](https://huggingface.co/sshleifer/mbart-large-cc25) :

```python
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("sshleifer/mbart-large-cc25")
```

---

Running this code yield an error :

>OSError: Model name 'sshleifer/mbart-large-cc25' was not found in tokenizers model name list (facebook/bart-large, facebook/bart-large-mnli, facebook/bart-large-cnn, facebook/bart-large-xsum). We assumed 'sshleifer/mbart-large-cc25' was a path, a model identifier, or url to a directory containing vocabulary files named ['vocab.json', 'merges.txt'] but couldn't find such vocabulary files at this path or url.

---

Which Tokenizer should I use with this model ?

@sshleifer

### PR DESCRIPTION
**Pull request by sshleifer**
TODO:

- [x] fix config (max_length=100)
- [x] split test_modeling_mbart.py to new file
- [x] decoder_input_ids should start with decoder_start_token_id
- [x] AutoTokenizer
- [x] test finetuning. Requires adding `--src_lang` and `--tgt_lang` clargs. 
- [x] Documentation
- [ ] Model Card
- [x] fix mbart tokenizer tests
- [x] tokenizer docs

### CODE REVIEW

File: tests/test_modeling_mbart.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    def model(self):
+        """Only load the model if needed."""
+        model = BartForConditionalGeneration.from_pretrained(self.checkpoint_name).to(torch_device)
+        if "cuda" in torch_device:
+            model = model.half()
+        return model
+
+
+@require_torch
+class MBartEnroIntegrationTest(MBartIntegrationTests):
```
Comment (sshleifer) [inline tests/test_modeling_mbart.py:36]: old logic copied to new file

File: tests/test_modeling_mbart.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        desired_max_length = 10
+        ids = self.tokenizer.prepare_translation_batch(
+            src_text, return_tensors=None, max_length=desired_max_length
+        ).input_ids[0]
+        self.assertEqual(ids[-2], 2)
+        self.assertEqual(ids[-1], EN_CODE)
+        self.assertEqual(len(ids), desired_max_length)
+
+
+class MBartCC25IntegrationTest(MBartIntegrationTests):
```
Comment (sshleifer) [inline tests/test_modeling_mbart.py:139]: New test

Comment (sshleifer) [reply tests/test_modeling_mbart.py:139]: This might be an unreasonable expectation.

File: examples/seq2seq/utils.py [hunk 3]
```diff
@@ -42,11 +42,11 @@ def encode_file(
     examples = []
     for text in tqdm(lns, desc=f"Tokenizing {data_path.name}"):
         tokenized = tokenizer(
-            [text],
+            [" " + text],
             max_length=max_length,
             padding="max_length" if pad_to_max_length else None,
             truncation=True,
-            add_prefix_space=True,
+            # add_prefix_space=True,
```
Comment (LysandreJik) [inline examples/seq2seq/utils.py:49]: Why this change?

File: src/transformers/tokenization_bart.py [hunk 4]
```diff
@@ -109,24 +109,33 @@ class MBartTokenizer(XLMRobertaTokenizer):
...
         self._additional_special_tokens = list(self.lang_code_to_id.keys())
+        self.reset_special_tokens()
+
+    def reset_special_tokens(self):
+
+        self.prefix_tokens = []
+        self.suffix_tokens = [self.eos_token_id, self.cur_lang_code]
 
     def build_inputs_with_special_tokens(self, token_ids_0, token_ids_1=None) -> List[int]:
         """Build model inputs from a sequence by appending eos_token_id."""
```
Comment (LysandreJik) [inline src/transformers/tokenization_bart.py:127]: Can the doc here be similar to the other tokenizers?

**Review by LysandreJik — COMMENTED** [review]
Could we test the MBart tokenizer? Or do you think it's too similar to the XLM-R tokenizer to deserve its own test suite?

File: examples/seq2seq/utils.py (hunk 3 above)
Comment (sshleifer) [reply examples/seq2seq/utils.py:49]: It now creates a warning, and I lazily haven't checked whether the outputs have changed.
I will resolve in this PR.

**Comment by sshleifer** [comment]
> Could we test the MBart tokenizer? Or do you think it's too similar to the XLM-R tokenizer to deserve its own test suite?

<3 that idea. Will do in this PR. Thanks for reminding me.

**Comment by sshleifer** [comment]
This does not need review at the moment. I need to fix the tests first.

File: examples/seq2seq/utils.py (hunk 3 above)
Comment (sshleifer) [reply examples/seq2seq/utils.py:49]: I can't reproduce the warning. going back to add_prefix_space=True.

File: src/transformers/tokenization_bart.py (hunk 4 above)
Comment (sshleifer) [reply src/transformers/tokenization_bart.py]: done

File: examples/longform-qa/eli5_app.py [hunk 5]
```diff
@@ -1,10 +1,10 @@
 import faiss
 import nlp
 import numpy as np
+import streamlit as st
```
Comment (sshleifer) [inline examples/longform-qa/eli5_app.py:4]: I updated setup.cfg to make this stop.

File: src/transformers/tokenization_bart.py [hunk 6]
```diff
@@ -109,24 +109,83 @@ class MBartTokenizer(XLMRobertaTokenizer):
...
+        Pairs of sequences are not the expected use case, but they will be handled without a separator.
+
+        Args:
+            token_ids_0 (:obj:`List[int]`):
+                List of IDs to which the special tokens will be added
+            token_ids_1 (:obj:`List[int]`, `optional`, defaults to :obj:`None`):
+                Optional second list of IDs for sequence pairs.
+
+        Returns:
+            :obj:`List[int]`: list of `input IDs <../glossary.html#input-ids>`__ with the appropriate special tokens.
```
Comment (LysandreJik) [inline src/transformers/tokenization_bart.py:145]: Very cool docstring

File: src/transformers/tokenization_bart.py [hunk 7]
```diff
@@ -135,44 +194,45 @@ def prepare_translation_batch(
         tgt_texts: Optional[List[str]] = None,
         tgt_lang: str = "ro_RO",
         max_length: Optional[int] = None,
-        pad_to_max_length: bool = True,
+        padding: str = "longest",
```
Comment (LysandreJik) [inline src/transformers/tokenization_bart.py:197]: This is a breaking change; is `prepare_translation_batch` user-facing? If so, can we keep backwards compatibility by still handling `pad_to_max_length` with a `FutureWarning`?

**Review by LysandreJik — APPROVED** [review]
This is very cool, love the tests.

**Comment by LysandreJik** [comment]
Oh, and you have the `build_doc` failing (warnings make the build fail) with the following error:

```
tokenization_bart.py:docstring of transformers.MBartTokenizer.build_inputs_with_special_tokens:4:Unexpected indentation.
```

File: src/transformers/tokenization_bart.py (hunk 7 above)
Comment (sshleifer) [reply src/transformers/tokenization_bart.py:198]: sure, good idea!

File: examples/seq2seq/utils.py [hunk 8]
```diff
@@ -42,11 +42,11 @@ def encode_file(
     examples = []
     for text in tqdm(lns, desc=f"Tokenizing {data_path.name}"):
         tokenized = tokenizer(
-            [" " + text],
+            [text],
```
Comment (patrickvonplaten) [inline examples/seq2seq/utils.py]: Is that left-over code from the T5 preprocessing?

File: src/transformers/tokenization_bart.py [hunk 9]
```diff
@@ -118,14 +118,32 @@ def __init__(self, *args, **kwargs):
...
-    def reset_special_tokens(self):
-
+    def reset_special_tokens(self) -> None:
+        """Reset the special tokens to the source lang setting. No prefix and suffix [eos, cur_lang_code]."""
         self.prefix_tokens = []
         self.suffix_tokens = [self.eos_token_id, self.cur_lang_code]
 
-    def build_inputs_with_special_tokens(self, token_ids_0, token_ids_1=None) -> List[int]:
-        """Build model inputs from a sequence by appending eos_token_id."""
+    def build_inputs_with_special_tokens(
```
Comment (patrickvonplaten) [inline src/transformers/tokenization_bart.py:128]: Nice docstring!

**Review by patrickvonplaten — APPROVED** [review]

File: examples/seq2seq/utils.py (hunk 8 above)
Comment (sshleifer) [reply examples/seq2seq/utils.py]: no it was a hack to replicate add_prefix_space=True. Does `T5Tokenizer` need that kwarg?

**Comment by Mehrad0711** [comment]
Hi Sam, 
Thanks for your great work to add MBART-cc25.
I've been following issues and threads regarding MBART generation problems for a while now.
Are those problems fixed in this PR or is the model still generating English?

**Comment by sshleifer** [comment]
The model is still generating english, but I'm not sure whether that's a bug. I can finetune it to achieve good scores on WMT en-ro, and nobody has replied to https://github.com/pytorch/fairseq/issues/2258, so I moved on.

**Comment by Mehrad0711** [comment]
Sad :(  I was really looking forward to trying it on language pairs that Marian does not support (e.g. English->Chinese, English->Arabic). Is there any way to get MBART to work without any fine-tuning?
The problem I'm facing is it always generates BOS tokens at the first decoding step even though the tgt_lang id is provided as the starting token. I tried forcing zero probability on BOS token at each step but then it generates gibberish text...

**Comment by nhatsmrt** [comment]
I don't know how the preprocessing is done in the original setup, but I found a potential problems in generating decoder_input_ids (i.e target sentence) for training. According to the documentation, it's supposed to have the format:
<language code> <tokens> <eos>
(which is different from the format of the input_ids), but the tokenizer currently generates:
<tokens> <eos> <language code>
For both source (input_ids) and target (decoder_input_ids). Maybe this is why you are unable to generate?

**Comment by sshleifer** [comment]
This was a bug, but is fixed on master if you use `prepare_translation_batch`.

For example
```python
from transformers import MBartTokenizer
model_name = 'facebook/mbart-large-cc25'
tok = MBartTokenizer.from_pretrained(model_name)
src_text = ['I think I fixed the tokenizer']
tgt_text = ['Cred că am rezolvat tokenizatorul']
batch = tok.prepare_translation_batch(src_text, tgt_texts=tgt_text)
batch.input_ids # ends in (eos, lang_code)
=> tensor([[    87,   5351,     87, 188347,     70,     47,   1098,  52825,      2, 250004]])
batch.decoder_input_ids # (lang_code, *tokens, eos)
=> tensor([[250020,  68523,   1362,    444, 102432,     18,     47,   1098, 164077, 202,      2]])
```

More info: mbart knows whether switches between src and tgt modes using the `set_lang` method: https://github.com/huggingface/transformers/blob/d6eab53058015483e9cbcbfee4bf900c3a8ab772/src/transformers/tokenization_bart.py#L186
